fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java

```diff
@@ -257,6 +283,210 @@ public PlanFragment visitPhysicalAggregate(
+        TupleDescriptor outputTupleDesc = tupleResult.getTupleDescriptor();
+        setSlotNullable(outputTupleDesc, finalSlots);
+
+        GroupingInfo groupingInfo = new GroupingInfo(ExpressionTranslator.translateGroupingType(repeat),
+                virtualTupleDesc, outputTupleDesc, preRepeatExprs);
+
+        // 2.create repeat node
+        // Replace the bitset of groupingsets with the bitset corresponding to slotId
+        List<Set<Integer>> repeatSlotIdList = genGroupingIdList(
+                repeat.useBitsetsToRepresentGroupingSets(), tupleResult.getSlotDescriptors());
+        Set<Integer> allSlotId = genAllSlotId(repeatSlotIdList);
+
+        RepeatNode repeatNode = new RepeatNode(context.nextPlanNodeId(),
+                inputPlanFragment.getPlanRoot(), groupingInfo, repeatSlotIdList,
+                allSlotId, repeat.getVirtualSlotValues());
+        repeatNode.setNumInstances(inputPlanFragment.getPlanRoot().getNumInstances());
+        inputPlanFragment.addPlanRoot(repeatNode);
+        inputPlanFragment.updateDataPartition(DataPartition.RANDOM);
+        return inputPlanFragment;
+    }
+
+    /**
+     * Generate outputSlot based on groupBy and output information.
+     * The generated outputSlot needs to guarantee the order,
+     * first slotReference and then virtualSlotReference.
+     *
+     * First collect the slot and then collect the virtualSlot.
+     * 1. Put the columns in groupBy into the slot
+     * 2. Put the columns in the output into the slot
+     * 3. Put slot and virtualSlot into finalSLots in turn.
+     *
+     * eg: select sum(k2) grouping(k1) from t1 group by grouping sets((k1));
+     * 1. slots:[k1]
+     * 2. slots:[k1, k2]
+     * 3. finalSlots: [k1, k2, GROUPING_PREFIX_k1]
+     */
+    private List<Slot> genOutputSlots(
+            List<SlotReference> groupBySlot,
+            List<NamedExpression> outputExpressionList,
+            List<Slot> virtualSlots) {
+
+        Map<Boolean, List<Slot>> allSlots = groupBySlot.stream()
+                .collect(Collectors.groupingBy(VirtualSlotReference.class::isInstance,
+                        LinkedHashMap::new, Collectors.toList()));
+        List<Slot> nonVirtualSlots = allSlots.get(false);
+        Map<Boolean, List<NamedExpression>> allOutput = outputExpressionList.stream()
+                .collect(Collectors.groupingBy(VirtualSlotReference.class::isInstance,
+                        LinkedHashMap::new, Collectors.toList()));
+        List<NamedExpression> nonVirtualOutput = allOutput.get(false);
+
+        nonVirtualSlots.addAll(nonVirtualOutput.stream()
+                .filter(e -> !nonVirtualSlots.contains(e))
+                .filter(SlotReference.class::isInstance)
+                .map(NamedExpression::toSlot)
+                .collect(Collectors.toSet()));
+        List<Slot> finalSlots = Lists.newArrayList();
+        finalSlots.addAll(nonVirtualSlots);
+        finalSlots.addAll(virtualSlots);
+        return finalSlots;
+    }
```

Comment:
output maybe not slot, but contains slot, and you need sort it, so getUsedSortedSlots is more reasonable. And primary method should move to the tailing of the class.

And this method should be simplified to
```java
private List<Slot> getUsedSortedSlots(List<Expression> groupByExpressions, List<NamedExpression> outputExpressions)
    Set <Slot> allSlots = ImmutableSet.builder()
        .addAll(groupByExpressions.stream()
            .flatMap(groupExpr -> groupExpr.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        .addAll(outputExpressions.stream()
            .flatMap(outputExpr -> outputExpr.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        ).build();

    Map<Boolean, Set<Slot>> whetherIsVirtualSlots = groupBySlot.stream()
                .collect(Collectors.groupingBy(VirtualSlotReference.class::isInstance,
                        LinkedHashMap::new, ImmutableSet.toImmutableSet()));

    return ImmutableList.Builder()
        .addAll(whetherIsVirtualSlots.get(false)) // non virtual slots
        .addAll(whetherIsVirtualSlots.get(true)) // virtual slots
        .build();
}
```